### PR TITLE
Fix #8918: Mergetool and difftool commands are now populated if paths…

### LINKED ISSUE #8918: Mergetool and difftool commands not populated if paths are manually configured
## Current behaviour

When you open Tools -> Settings -> Git -> Config and specify the mergetool or difftool (the behavior is the same for both) the corresponding command configuration is only auto-populated if your browse to the executable. It is not populated if you manually configure a path, e.g. copy-paste the path to kdiff3.exe into the path field.

## Expected behaviour

The command fields (both for mergetool and difftool) are auto-populated if empty not just when you browse to the tool executable and select it like that but also if you add the path manually directly to the field.

## Steps to reproduce

1. Go to Tools -> Settings -> Git -> Config.
2. Make sure mergetool is not configured, like it is with a fresh install. I.e. "Path to mergetool" and "Mergetool command" is empty (the "Mergetool" dropdown being unselected is not necessary).
3. Click "Browse" to select the mergetool exe.
4. Observe "Mergetool command" being populated. This is an expected behavior.
5. Do 2. again.
6. This time, write the path into the "Path to mergetool" field, don't browse.
7. Observe "Mergetool command" being left empty.

It's the same for difftool.

## Screenshots

Speaking about this UI:

image

## Did this work in previous version of GitExtensions

Don't know.

## Environment

- Git Extensions 3.4.3.9999
- Build d4b0f48bbf3e39a71f5d7ff5231be5678d4aa0f1
- Git 2.30.1.windows.1
- Microsoft Windows NT 10.0.19042.0
- .NET Framework 4.8.4300.0
- DPI 96dpi (no scaling)

## Diagnostics

Doesn't seem to be applicable.

### PR DESCRIPTION
**Pull request by blazejszuca**
… are manually configured.

Fixes #8918

## Proposed changes

- In configuration section of git tools: Mergetool and difftool commands are populated if paths to the tools are changed manually without need of pressing "suggest" button.

## Test methodology 

- manual tests

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Review by gerhardol — CHANGES_REQUESTED** [review]
If the tool has special support in GE, this is fine.
However, if the toolname is unknown to GE, the command is just erased which is quite unexpected. 

One way to handle this is to run the "auto suggest" only for known tools.
A little hacky though, need a good comment in the _TextChanged() methods.

File: GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.Designer.cs [hunk 1]
```diff
@@ -170,6 +170,8 @@ private void InitializeComponent()
             this.txtDiffToolPath.Location = new System.Drawing.Point(129, 200);
             this.txtDiffToolPath.Name = "txtDiffToolPath";
             this.txtDiffToolPath.Size = new System.Drawing.Size(902, 20);
+            this.txtDiffToolPath.TabIndex = 17;
```
[the remark this answers is not in the document]
Comment (blazejszuca) [reply GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.Designer.cs]: removed

File: GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.Designer.cs [hunk 2]
```diff
@@ -288,6 +290,8 @@ private void InitializeComponent()
             this.txtMergeToolPath.Location = new System.Drawing.Point(129, 111);
             this.txtMergeToolPath.Name = "txtMergeToolPath";
             this.txtMergeToolPath.Size = new System.Drawing.Size(902, 20);
+            this.txtMergeToolPath.TabIndex = 9;
```
[the remark this answers is not in the document]
Comment (blazejszuca) [reply GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.Designer.cs]: removed

File: GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs [hunk 3]
```diff
@@ -332,5 +332,21 @@ private void ConfigureEncoding_Click(object sender, EventArgs e)
                 CommonLogic.FillEncodings(Global_FilesEncoding);
             }
         }
+
+        private void txtMergeToolPath_TextChanged(object sender, EventArgs e)
```
[the remark this answers is not in the document]
Comment (blazejszuca) [reply GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs]: Moved below _LostFocus method.

**Comment by blazejszuca** [comment]
> If the tool has special support in GE, this is fine. However, if the toolname is unknown to GE, the command is just erased which is quite unexpected.
> 
> One way to handle this is to run the "auto suggest" only for known tools. A little hacky though, need a good comment in the _TextChanged() methods.

Totally agree - missed that point. Now events first check if tool is known and alter the command only for known tools.

File: GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs [hunk 4]
```diff
@@ -250,6 +250,32 @@ private void txtDiffMergeToolPath_LostFocus(object sender, EventArgs e)
...
+            }
+        }
+
+        private void txtDiffToolPath_TextChanged(object sender, EventArgs e)
+        {
+            if (txtDiffToolPath.Focused && sender == txtDiffToolPath)
+            {
+                // we only want to suggest command for known tool types
+                var toolName = _NO_TRANSLATE_cboDiffTool.Text;
+                if (!string.IsNullOrWhiteSpace(toolName))
```
Comment (gerhardol) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs:272]: Still an issue, should be checked for difftools known by GE. If you add your own tool and call it like `kdiff3_updated` then the cmd is erased at all changes.

Comment (blazejszuca) [reply GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs]: Ok. I added verification for built-in tools. Please note that "Suggest" button doesn't care about such verification.

Comment (gerhardol) [reply GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs]: > Please note that "Suggest" button doesn't care about such verification.

That is an explicit request to Suggest at least, you can avoid it. (If the tool is known in the gitconfig, the full command is filled in when typing the name so the default use case is supported).
You could argue that the check should be in `SuggestDiffToolCommand()` instead, but I believe that the `Suggest` button should be disabled in this case instead. Some more logic though.

The concept that GE knows about some tools improves usability some but makes the handling more complex...
(If path is quouted, the path is quoted again in command)

**Review by gerhardol — APPROVED** [review]
Plan to merge when other reviewers approve or in about four days.

**Comment by gerhardol** [comment]
cla need to be signed (see contributing), normally easiest in a separate PR

**Comment by blazejszuca** [comment]
> cla need to be signed (see contributing), normally easiest in a separate PR

Signed through PR #9781

File: GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.Designer.cs [hunk 5]
```diff
@@ -170,6 +170,7 @@ private void InitializeComponent()
             this.txtDiffToolPath.Location = new System.Drawing.Point(129, 200);
             this.txtDiffToolPath.Name = "txtDiffToolPath";
             this.txtDiffToolPath.Size = new System.Drawing.Size(902, 20);
+            this.txtDiffToolPath.TextChanged += new System.EventHandler(this.txtDiffToolPath_TextChanged);
             this.txtDiffToolPath.LostFocus += new System.EventHandler(this.txtDiffMergeToolPath_LostFocus);
```
Comment (RussKie) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.Designer.cs:174]: nit, the designer orders statements in A-Z order
```suggestion
            this.txtDiffToolPath.LostFocus += new System.EventHandler(this.txtDiffMergeToolPath_LostFocus);
            this.txtDiffToolPath.TextChanged += new System.EventHandler(this.txtDiffToolPath_TextChanged);
```

File: GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.Designer.cs [hunk 6]
```diff
@@ -288,6 +289,7 @@ private void InitializeComponent()
             this.txtMergeToolPath.Location = new System.Drawing.Point(129, 111);
             this.txtMergeToolPath.Name = "txtMergeToolPath";
             this.txtMergeToolPath.Size = new System.Drawing.Size(902, 20);
+            this.txtMergeToolPath.TextChanged += new System.EventHandler(this.txtMergeToolPath_TextChanged);
             this.txtMergeToolPath.LostFocus += new System.EventHandler(this.txtDiffMergeToolPath_LostFocus);
```
Comment (RussKie) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.Designer.cs:293]:
```suggestion
            this.txtMergeToolPath.LostFocus += new System.EventHandler(this.txtDiffMergeToolPath_LostFocus);
            this.txtMergeToolPath.TextChanged += new System.EventHandler(this.txtMergeToolPath_TextChanged);
```

File: GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs [hunk 7]
```diff
@@ -250,6 +250,32 @@ private void txtDiffMergeToolPath_LostFocus(object sender, EventArgs e)
             }
         }
 
+        private void txtMergeToolPath_TextChanged(object sender, EventArgs e)
+        {
+            if (txtMergeToolPath.Focused && sender == txtMergeToolPath)
```
Comment (RussKie) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs:255]: Please invert to reduce nesting

File: GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs [hunk 8]
```diff
@@ -250,6 +250,32 @@ private void txtDiffMergeToolPath_LostFocus(object sender, EventArgs e)
...
+                if (!string.IsNullOrWhiteSpace(toolName) && RegisteredDiffMergeTools.All(DiffMergeToolType.Merge).Contains(toolName))
+                {
+                    SuggestMergeToolCommand();
+                }
+            }
+        }
+
+        private void txtDiffToolPath_TextChanged(object sender, EventArgs e)
+        {
+            if (txtDiffToolPath.Focused && sender == txtDiffToolPath)
```
Comment (RussKie) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs:268]: ditto

**Review by RussKie — COMMENTED** [review]
LGTM

File: GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.Designer.cs (hunk 5 above)
Comment (blazejszuca) [reply GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.Designer.cs]: You are right - strange enough I didn't add it manually. Reordered.

File: GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.Designer.cs (hunk 6 above)
Comment (blazejszuca) [reply GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.Designer.cs]: Reordered.

File: GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs (hunk 7 above)
Comment (blazejszuca) [reply GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs]: Nesting reduced.

File: GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs (hunk 8 above)
Comment (blazejszuca) [reply GitUI/CommandsDialogs/SettingsDialog/Pages/GitConfigSettingsPage.cs]: Nesting reduced.

**Comment by gerhardol** [comment]
squash merging in a day

**Comment by RussKie** [comment]
👍

**Pull request merged**